stubs/openpyxl/openpyxl/styles/numbers.pyi

```diff
@@ -51,12 +51,18 @@ LOCALE_GROUP: Final = r"\[(?!hh?\]|mm?\]|ss?\])[^\]]*\]"
 STRIP_RE: Final[Pattern[str]]
 TIMEDELTA_RE: Final[Pattern[str]]
 
-def is_date_format(fmt) -> bool: ...
-def is_timedelta_format(fmt): ...
-def is_datetime(fmt): ...
-def is_builtin(fmt): ...
+def is_date_format(fmt: str | None) -> TypeGuard[str]: ...
+def is_timedelta_format(fmt: str | None) -> TypeGuard[str]: ...
```

Comment:
I'm unsure whether we can really use type guard here, as these do more than just check the type. Maybe it would make more sense to use overloads?

```suggestion
@overload
def is_date_format(fmt: None) -> None: ...
@overload
def is_date_format(fmt: str) -> bool: ...
@overload
def is_timedelta_format(fmt: None) -> None: ...
@overload
def is_timedelta_format(fmt: str) -> bool: ...
```

Cc @erictraut as resident type guard expert.

Implementation for reference: https://foss.heptapod.net/openpyxl/openpyxl/-/blob/branch/3.1/openpyxl/styles/numbers.py

Reply:
My understanding of `TypeGuard` is that if the function returns `True` then the parameter is necessarily of the guarded type. Which is the case for both here: `fmt=None` will imediatly return `False`. `fmt=""` may or may not return `True`.
I don't see anything in https://peps.python.org/pep-0647 suggesting that returning `False` has to narrow.

That being said, mypy seems to agree here, but pyright/pylance seems to ignore the typeguard, so I'll wait for an answer:
![image](https://github.com/python/typeshed/assets/1350584/416c7a22-fb83-41e1-b233-b66ca730093a)
![image](https://github.com/python/typeshed/assets/1350584/e69ee1bb-abc4-4009-9a7b-91dae25c5990)
![image](https://github.com/python/typeshed/assets/1350584/5f41df3a-3ecd-45c8-884a-b8f30e765f0b)

Reply:
This looks to me like a fine use of `TypeGuard`. The overload approach is correct from a type perspective, but it won't result in type narrowing when the function is used in a conditional statement, so `TypeGuard` is probably more appropriate here than an overload.

Both pyright and mypy honor the type guard in this test code.

```python
from typing import TypeGuard


def is_date_format(fmt: None | str) -> TypeGuard[str]:
    ...

def func(test: str | None):
    if is_date_format(test):
        reveal_type(test)  # str
    else:
        reveal_type(test)  # str | None
```